// @elas_co/bitcoints/lib/op-code.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpCode = void 0;
/*
 * OpCode
 * ======
 *
 * An opCode is one of the operations in the bitcoin scripting language. Each
 * operation is just a number from 0-255, and it has a corresponding string,
 * e.g. "OP_RETURN", which comes from the name of that constant in the bitcoind
 * source code. The way you probably want to use this is with
 * new OpCode(str).toNumber() or new OpCode(num).toString()
 */
const struct_1 = require("./struct");
var OpCodeValue;
(function (OpCodeValue) {
    // push value
    OpCodeValue[OpCodeValue["OP_FALSE"] = 0] = "OP_FALSE";
    OpCodeValue[OpCodeValue["OP_0"] = 0] = "OP_0";
    OpCodeValue[OpCodeValue["OP_PUSHDATA1"] = 76] = "OP_PUSHDATA1";
    OpCodeValue[OpCodeValue["OP_PUSHDATA2"] = 77] = "OP_PUSHDATA2";
    OpCodeValue[OpCodeValue["OP_PUSHDATA4"] = 78] = "OP_PUSHDATA4";
    OpCodeValue[OpCodeValue["OP_1NEGATE"] = 79] = "OP_1NEGATE";
    OpCodeValue[OpCodeValue["OP_RESERVED"] = 80] = "OP_RESERVED";
    OpCodeValue[OpCodeValue["OP_TRUE"] = 81] = "OP_TRUE";
    OpCodeValue[OpCodeValue["OP_1"] = 81] = "OP_1";
    OpCodeValue[OpCodeValue["OP_2"] = 82] = "OP_2";
    OpCodeValue[OpCodeValue["OP_3"] = 83] = "OP_3";
    OpCodeValue[OpCodeValue["OP_4"] = 84] = "OP_4";
    OpCodeValue[OpCodeValue["OP_5"] = 85] = "OP_5";
    OpCodeValue[OpCodeValue["OP_6"] = 86] = "OP_6";
    OpCodeValue[OpCodeValue["OP_7"] = 87] = "OP_7";
    OpCodeValue[OpCodeValue["OP_8"] = 88] = "OP_8";
    OpCodeValue[OpCodeValue["OP_9"] = 89] = "OP_9";
    OpCodeValue[OpCodeValue["OP_10"] = 90] = "OP_10";
    OpCodeValue[OpCodeValue["OP_11"] = 91] = "OP_11";
    OpCodeValue[OpCodeValue["OP_12"] = 92] = "OP_12";
    OpCodeValue[OpCodeValue["OP_13"] = 93] = "OP_13";
    OpCodeValue[OpCodeValue["OP_14"] = 94] = "OP_14";
    OpCodeValue[OpCodeValue["OP_15"] = 95] = "OP_15";
    OpCodeValue[OpCodeValue["OP_16"] = 96] = "OP_16";
    // control
    OpCodeValue[OpCodeValue["OP_NOP"] = 97] = "OP_NOP";
    OpCodeValue[OpCodeValue["OP_VER"] = 98] = "OP_VER";
    OpCodeValue[OpCodeValue["OP_IF"] = 99] = "OP_IF";
    OpCodeValue[OpCodeValue["OP_NOTIF"] = 100] = "OP_NOTIF";
    OpCodeValue[OpCodeValue["OP_VERIF"] = 101] = "OP_VERIF";
    OpCodeValue[OpCodeValue["OP_VERNOTIF"] = 102] = "OP_VERNOTIF";
    OpCodeValue[OpCodeValue["OP_ELSE"] = 103] = "OP_ELSE";
    OpCodeValue[OpCodeValue["OP_ENDIF"] = 104] = "OP_ENDIF";
    OpCodeValue[OpCodeValue["OP_VERIFY"] = 105] = "OP_VERIFY";
    OpCodeValue[OpCodeValue["OP_RETURN"] = 106] = "OP_RETURN";
    // stack ops
    OpCodeValue[OpCodeValue["OP_TOALTSTACK"] = 107] = "OP_TOALTSTACK";
    OpCodeValue[OpCodeValue["OP_FROMALTSTACK"] = 108] = "OP_FROMALTSTACK";
    OpCodeValue[OpCodeValue["OP_2DROP"] = 109] = "OP_2DROP";
    OpCodeValue[OpCodeValue["OP_2DUP"] = 110] = "OP_2DUP";
    OpCodeValue[OpCodeValue["OP_3DUP"] = 111] = "OP_3DUP";
    OpCodeValue[OpCodeValue["OP_2OVER"] = 112] = "OP_2OVER";
    OpCodeValue[OpCodeValue["OP_2ROT"] = 113] = "OP_2ROT";
    OpCodeValue[OpCodeValue["OP_2SWAP"] = 114] = "OP_2SWAP";
    OpCodeValue[OpCodeValue["OP_IFDUP"] = 115] = "OP_IFDUP";
    OpCodeValue[OpCodeValue["OP_DEPTH"] = 116] = "OP_DEPTH";
    OpCodeValue[OpCodeValue["OP_DROP"] = 117] = "OP_DROP";
    OpCodeValue[OpCodeValue["OP_DUP"] = 118] = "OP_DUP";
    OpCodeValue[OpCodeValue["OP_NIP"] = 119] = "OP_NIP";
    OpCodeValue[OpCodeValue["OP_OVER"] = 120] = "OP_OVER";
    OpCodeValue[OpCodeValue["OP_PICK"] = 121] = "OP_PICK";
    OpCodeValue[OpCodeValue["OP_ROLL"] = 122] = "OP_ROLL";
    OpCodeValue[OpCodeValue["OP_ROT"] = 123] = "OP_ROT";
    OpCodeValue[OpCodeValue["OP_SWAP"] = 124] = "OP_SWAP";
    OpCodeValue[OpCodeValue["OP_TUCK"] = 125] = "OP_TUCK";
    // data manipulation ops
    OpCodeValue[OpCodeValue["OP_CAT"] = 126] = "OP_CAT";
    OpCodeValue[OpCodeValue["OP_SUBSTR"] = 127] = "OP_SUBSTR";
    OpCodeValue[OpCodeValue["OP_SPLIT"] = 127] = "OP_SPLIT";
    OpCodeValue[OpCodeValue["OP_LEFT"] = 128] = "OP_LEFT";
    OpCodeValue[OpCodeValue["OP_NUM2BIN"] = 128] = "OP_NUM2BIN";
    OpCodeValue[OpCodeValue["OP_RIGHT"] = 129] = "OP_RIGHT";
    OpCodeValue[OpCodeValue["OP_BIN2NUM"] = 129] = "OP_BIN2NUM";
    OpCodeValue[OpCodeValue["OP_SIZE"] = 130] = "OP_SIZE";
    // bit logic
    OpCodeValue[OpCodeValue["OP_INVERT"] = 131] = "OP_INVERT";
    OpCodeValue[OpCodeValue["OP_AND"] = 132] = "OP_AND";
    OpCodeValue[OpCodeValue["OP_OR"] = 133] = "OP_OR";
    OpCodeValue[OpCodeValue["OP_XOR"] = 134] = "OP_XOR";
    OpCodeValue[OpCodeValue["OP_EQUAL"] = 135] = "OP_EQUAL";
    OpCodeValue[OpCodeValue["OP_EQUALVERIFY"] = 136] = "OP_EQUALVERIFY";
    OpCodeValue[OpCodeValue["OP_RESERVED1"] = 137] = "OP_RESERVED1";
    OpCodeValue[OpCodeValue["OP_RESERVED2"] = 138] = "OP_RESERVED2";
    // numeric
    OpCodeValue[OpCodeValue["OP_1ADD"] = 139] = "OP_1ADD";
    OpCodeValue[OpCodeValue["OP_1SUB"] = 140] = "OP_1SUB";
    OpCodeValue[OpCodeValue["OP_2MUL"] = 141] = "OP_2MUL";
    OpCodeValue[OpCodeValue["OP_2DIV"] = 142] = "OP_2DIV";
    OpCodeValue[OpCodeValue["OP_NEGATE"] = 143] = "OP_NEGATE";
    OpCodeValue[OpCodeValue["OP_ABS"] = 144] = "OP_ABS";
    OpCodeValue[OpCodeValue["OP_NOT"] = 145] = "OP_NOT";
    OpCodeValue[OpCodeValue["OP_0NOTEQUAL"] = 146] = "OP_0NOTEQUAL";
    OpCodeValue[OpCodeValue["OP_ADD"] = 147] = "OP_ADD";
    OpCodeValue[OpCodeValue["OP_SUB"] = 148] = "OP_SUB";
    OpCodeValue[OpCodeValue["OP_MUL"] = 149] = "OP_MUL";
    OpCodeValue[OpCodeValue["OP_DIV"] = 150] = "OP_DIV";
    OpCodeValue[OpCodeValue["OP_MOD"] = 151] = "OP_MOD";
    OpCodeValue[OpCodeValue["OP_LSHIFT"] = 152] = "OP_LSHIFT";
    OpCodeValue[OpCodeValue["OP_RSHIFT"] = 153] = "OP_RSHIFT";
    OpCodeValue[OpCodeValue["OP_BOOLAND"] = 154] = "OP_BOOLAND";
    OpCodeValue[OpCodeValue["OP_BOOLOR"] = 155] = "OP_BOOLOR";
    OpCodeValue[OpCodeValue["OP_NUMEQUAL"] = 156] = "OP_NUMEQUAL";
    OpCodeValue[OpCodeValue["OP_NUMEQUALVERIFY"] = 157] = "OP_NUMEQUALVERIFY";
    OpCodeValue[OpCodeValue["OP_NUMNOTEQUAL"] = 158] = "OP_NUMNOTEQUAL";
    OpCodeValue[OpCodeValue["OP_LESSTHAN"] = 159] = "OP_LESSTHAN";
    OpCodeValue[OpCodeValue["OP_GREATERTHAN"] = 160] = "OP_GREATERTHAN";
    OpCodeValue[OpCodeValue["OP_LESSTHANOREQUAL"] = 161] = "OP_LESSTHANOREQUAL";
    OpCodeValue[OpCodeValue["OP_GREATERTHANOREQUAL"] = 162] = "OP_GREATERTHANOREQUAL";
    OpCodeValue[OpCodeValue["OP_MIN"] = 163] = "OP_MIN";
    OpCodeValue[OpCodeValue["OP_MAX"] = 164] = "OP_MAX";
    OpCodeValue[OpCodeValue["OP_WITHIN"] = 165] = "OP_WITHIN";
    // crypto
    OpCodeValue[OpCodeValue["OP_RIPEMD160"] = 166] = "OP_RIPEMD160";
    OpCodeValue[OpCodeValue["OP_SHA1"] = 167] = "OP_SHA1";
    OpCodeValue[OpCodeValue["OP_SHA256"] = 168] = "OP_SHA256";
    OpCodeValue[OpCodeValue["OP_HASH160"] = 169] = "OP_HASH160";
    OpCodeValue[OpCodeValue["OP_HASH256"] = 170] = "OP_HASH256";
    OpCodeValue[OpCodeValue["OP_CODESEPARATOR"] = 171] = "OP_CODESEPARATOR";
    OpCodeValue[OpCodeValue["OP_CHECKSIG"] = 172] = "OP_CHECKSIG";
    OpCodeValue[OpCodeValue["OP_CHECKSIGVERIFY"] = 173] = "OP_CHECKSIGVERIFY";
    OpCodeValue[OpCodeValue["OP_CHECKMULTISIG"] = 174] = "OP_CHECKMULTISIG";
    OpCodeValue[OpCodeValue["OP_CHECKMULTISIGVERIFY"] = 175] = "OP_CHECKMULTISIGVERIFY";
    // expansion
    OpCodeValue[OpCodeValue["OP_NOP1"] = 176] = "OP_NOP1";
    OpCodeValue[OpCodeValue["OP_NOP2"] = 177] = "OP_NOP2";
    OpCodeValue[OpCodeValue["OP_CHECKLOCKTIMEVERIFY"] = 177] = "OP_CHECKLOCKTIMEVERIFY";
    OpCodeValue[OpCodeValue["OP_NOP3"] = 178] = "OP_NOP3";
    OpCodeValue[OpCodeValue["OP_CHECKSEQUENCEVERIFY"] = 178] = "OP_CHECKSEQUENCEVERIFY";
    OpCodeValue[OpCodeValue["OP_NOP4"] = 179] = "OP_NOP4";
    OpCodeValue[OpCodeValue["OP_NOP5"] = 180] = "OP_NOP5";
    OpCodeValue[OpCodeValue["OP_NOP6"] = 181] = "OP_NOP6";
    OpCodeValue[OpCodeValue["OP_NOP7"] = 182] = "OP_NOP7";
    OpCodeValue[OpCodeValue["OP_NOP8"] = 183] = "OP_NOP8";
    OpCodeValue[OpCodeValue["OP_NOP9"] = 184] = "OP_NOP9";
    OpCodeValue[OpCodeValue["OP_NOP10"] = 185] = "OP_NOP10";
    // template matching params
    OpCodeValue[OpCodeValue["OP_SMALLDATA"] = 249] = "OP_SMALLDATA";
    OpCodeValue[OpCodeValue["OP_SMALLINTEGER"] = 250] = "OP_SMALLINTEGER";
    OpCodeValue[OpCodeValue["OP_PUBKEYS"] = 251] = "OP_PUBKEYS";
    OpCodeValue[OpCodeValue["OP_PUBKEYHASH"] = 253] = "OP_PUBKEYHASH";
    OpCodeValue[OpCodeValue["OP_PUBKEY"] = 254] = "OP_PUBKEY";
    OpCodeValue[OpCodeValue["OP_INVALIDOPCODE"] = 255] = "OP_INVALIDOPCODE";
})(OpCodeValue || (OpCodeValue = {}));
class _OpCode extends struct_1.Struct {
    constructor(num) {
        super({ num });
    }
    fromNumber(num) {
        this.num = num;
        return this;
    }
    static fromNumber(num) {
        return new this().fromNumber(num);
    }
    toNumber() {
        return this.num;
    }
    fromString(str) {
        const num = OpCodeValue[str];
        if (num === undefined) {
            throw new Error('Invalid opCodeStr');
        }
        this.num = num;
        return this;
    }
    toString() {
        var _a;
        const str = _OpCode.str[this.num];
        // console.log(str || this.num)
        if (str === undefined) {
            if (this.num > 0 && this.num < exports.OpCode.OP_PUSHDATA1) {
                return (_a = this === null || this === void 0 ? void 0 : this.num) === null || _a === void 0 ? void 0 : _a.toString();
            }
            return 'OP_UNKNOWN'; // str //undefined // throw new Error('OpCode does not have a string representation: ' + this.num)
        }
        return str;
    }
}
_OpCode.str = {};
for (const opCodeStr in OpCodeValue) {
    _OpCode[opCodeStr] = OpCodeValue[opCodeStr];
    if (Object.prototype.hasOwnProperty.call(OpCodeValue, opCodeStr)) {
        _OpCode.str[OpCodeValue[opCodeStr]] = opCodeStr;
    }
}
exports.OpCode = _OpCode;
