// @elas_co/bitcoints/lib/tx-out.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxOut = void 0;
/**
 * Transaction Output
 * ==================
 *
 * An output to a transaction. The way you normally want to make one is with
 * new TxOut(valueBn, script) (i.e., just as with TxIn, you can leave out the
 * scriptVi, since it can be computed automatically.
 */
const bn_1 = require("./bn");
const bw_1 = require("./bw");
const script_1 = require("./script");
const struct_1 = require("./struct");
const var_int_1 = require("./var-int");
class TxOut extends struct_1.Struct {
    constructor(valueBn, scriptVi, script) {
        super({ valueBn, scriptVi, script });
    }
    setScript(script) {
        this.scriptVi = var_int_1.VarInt.fromNumber(script.toBuffer().length);
        this.script = script;
        return this;
    }
    fromProperties(valueBn, script) {
        this.fromObject({ valueBn });
        this.setScript(script);
        return this;
    }
    static fromProperties(valueBn, script) {
        return new this().fromProperties(valueBn, script);
    }
    fromJSON(json) {
        this.fromObject({
            valueBn: new bn_1.Bn().fromJSON(json.valueBn),
            scriptVi: new var_int_1.VarInt().fromJSON(json.scriptVi),
            script: new script_1.Script().fromJSON(json.script),
        });
        return this;
    }
    toJSON() {
        return {
            valueBn: this.valueBn.toJSON(),
            scriptVi: this.scriptVi.toJSON(),
            script: this.script.toJSON(),
        };
    }
    fromBr(br) {
        this.valueBn = br.readUInt64LEBn();
        this.scriptVi = var_int_1.VarInt.fromNumber(br.readVarIntNum());
        this.script = new script_1.Script().fromBuffer(br.read(this.scriptVi.toNumber()));
        return this;
    }
    toBw(bw) {
        if (!bw) {
            bw = new bw_1.Bw();
        }
        bw.writeUInt64LEBn(this.valueBn);
        bw.write(this.scriptVi.buf);
        bw.write(this.script.toBuffer());
        return bw;
    }
}
exports.TxOut = TxOut;
