// @elas_co/bitcoints/lib/var-int.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VarInt = void 0;
const br_1 = require("./br");
const bw_1 = require("./bw");
const struct_1 = require("./struct");
class VarInt extends struct_1.Struct {
    constructor(buf) {
        super({ buf });
    }
    fromJSON(json) {
        this.fromObject({
            buf: Buffer.from(json, 'hex'),
        });
        return this;
    }
    toJSON() {
        return this.buf.toString('hex');
    }
    fromBuffer(buf) {
        this.buf = buf;
        return this;
    }
    fromBr(br) {
        this.buf = br.readVarIntBuf();
        return this;
    }
    fromBn(bn) {
        this.buf = new bw_1.Bw().writeVarIntBn(bn).toBuffer();
        return this;
    }
    static fromBn(bn) {
        return new this().fromBn(bn);
    }
    fromNumber(num) {
        this.buf = new bw_1.Bw().writeVarIntNum(num).toBuffer();
        return this;
    }
    static fromNumber(num) {
        return new this().fromNumber(num);
    }
    toBuffer() {
        return this.buf;
    }
    toBn() {
        return new br_1.Br(this.buf).readVarIntBn();
    }
    toNumber() {
        return new br_1.Br(this.buf).readVarIntNum();
    }
}
exports.VarInt = VarInt;
