// @elas_co/bitcoints/lib/workers-result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkersResult = void 0;
const bw_1 = require("./bw");
const struct_1 = require("./struct");
class WorkersResult extends struct_1.Struct {
    constructor(resbuf, isError, id) {
        super({ resbuf, isError, id });
    }
    fromResult(result, id) {
        if (Buffer.isBuffer(result)) {
            this.resbuf = result;
        }
        else if (result.toFastBuffer) {
            this.resbuf = result.toFastBuffer();
        }
        else {
            this.resbuf = Buffer.from(JSON.stringify(result));
        }
        this.isError = false;
        this.id = id;
        return this;
    }
    static fromResult(result, id) {
        return new this().fromResult(result, id);
    }
    fromError(error, id) {
        this.resbuf = Buffer.from(JSON.stringify(error.message));
        this.isError = true;
        this.id = id;
        return this;
    }
    toBw(bw) {
        if (!bw) {
            bw = new bw_1.Bw();
        }
        bw.writeVarIntNum(this.resbuf.length);
        bw.write(this.resbuf);
        bw.writeUInt8(Number(this.isError));
        bw.writeVarIntNum(this.id);
        return bw;
    }
    fromBr(br) {
        const resbuflen = br.readVarIntNum();
        this.resbuf = br.read(resbuflen);
        this.isError = Boolean(br.readUInt8());
        this.id = br.readVarIntNum();
        return this;
    }
}
exports.WorkersResult = WorkersResult;
